ktor-client/ktor-client-features/ktor-client-tracing/common/src/TracingWrapper.kt

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.features.tracing
+
+import io.ktor.client.*
+import io.ktor.client.engine.*
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.http.cio.websocket.*
+import io.ktor.util.*
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+
+/**
+ * Tracing wrapper that wraps [HttpClientEngineFactory] and creates [EngineWithTracer] instead of original engine.
+ */
+class TracingWrapper<T : HttpClientEngineConfig>(
+    private val delegate: HttpClientEngineFactory<T>,
+    private val tracer: Tracer
+) : HttpClientEngineFactory<T> {
+    override fun create(block: T.() -> Unit): HttpClientEngine {
+        val engine = delegate.create(block)
+        return EngineWithTracer(engine, tracer)
+    }
+}
+
+/**
+ * Engine with tracer that wraps request execution into tracing functionality. Calls correspondent [tracer] method on
+ * every important processing event.
+ */
+internal class EngineWithTracer(private val delegate: HttpClientEngine, private val tracer: Tracer) :
+    HttpClientEngine by delegate {
```

Comment:
Could you tell me why we need delegation here?

Reply:
Because we delegate `config`, `coroutineContext`, `dispatcher` and perhaps some other methods to the delegate.